// node-jpa/src/driver/mysql/Columns.js
module.exports = {
  int(field, isPrimaryKey, auto = true) {
    if (isPrimaryKey && auto) {
      return `\`${field.name}\` INT(11) NOT NULL AUTO_INCREMENT`
    } else {
      return `\`${field.name}\` INT(11) NOT NULL DEFAULT 0`
    }
  },
  string(field, isPrimaryKey) {
    let len = field.len || 255
    if (isPrimaryKey) {
      return `\`${field.name}\` VARCHAR(${len}) COLLATE utf8mb4_unicode_ci NOT NULL`
    } else {
      return `\`${field.name}\` VARCHAR(${len}) COLLATE utf8mb4_unicode_ci DEFAULT NULL`
    }
  },
  text(field) {
    return `\`${field.name}\` TEXT COLLATE utf8mb4_unicode_ci`
  },
  timestamp(field) {
    return `\`${field.name}\` TIMESTAMP NULL DEFAULT NULL`
  },
  decimal(field) {
    let m = field.m || 10
    let d = field.d || 0
    return `\`${field.name}\` DECIMAL(${m}, ${d}) NOT NULL DEFAULT 0`
  },
  float(field) {
    let m = field.m || 10
    let d = field.d || 0
    return `\`${field.name}\` FLOAT(${m}, ${d}) NOT NULL DEFAULT 0`
  },
  double(field) {
    let m = field.m || 10
    let d = field.d || 0
    return `\`${field.name}\` DOUBLE(${m}, ${d}) NOT NULL DEFAULT 0`
  }
}
