// node-jpa/src/driver/mysql/MysqlConnection.js
const Connection = require('../Connection')

module.exports = class MysqlConnection extends Connection {
  constructor(instance) {
    super()
    this._instance = instance
    this._timestamp = Date.now()
  }

  get timestamp() {
    return this._timestamp
  }

  query(sql, values) {
    return new Promise((resolve, reject) => {
      this._instance.query(sql, values, (err, result) => {
        err ? reject(err) : resolve(result)
      })
    })
  }

  beginTransaction() {
    return new Promise((resolve, reject) => {
      this._instance.beginTransaction(err => {
        err ? reject(err) : resolve()
      })
    })
  }

  commit() {
    return new Promise((resolve, reject) => {
      this._instance.commit(err => {
        err ? reject(err) : resolve()
      })
    })
  }

  rollback() {
    return new Promise((resolve, reject) => {
      this._instance.rollback(err => {
        err ? reject(err) : resolve()
      })
    })
  }

  release() {
    this._instance.release()
  }

  destroy() {
    this._instance.destroy()
  }
}
